### ISSUE TITLE
Support for Homekit accessory airpurifier

### ISSUE THREAD

**Issue by EjvindHald** [arachnetech/homebridge-mqttthing#185]
Hi guys.

Thank you for this really cool stuff you are providing. So far, everything I have used works seamlessly. Especially, I am pleased with the target and current state for some of the accessories are transparent and can be used for very fine grained two way control of my hardware and the Homekit IOS app.

If you would consider to add support for Homekit accessory type airpurifier, that would be great.

Thanks.
Ejvind

**Comment by arachnetech** [comment]
Sorry for my slow response. That certainly looks possible. I’ll add it to my list.

**Comment by EjvindHald** [comment]
👍

**Comment by tobekas** [comment]
Hi. I just scanned the specification for AirPurifier roughly and it seems to be feasible. There are, however, some options for linking other services (such as Fan or Slat), which can then be controlled separately if necessary. I would have to take a rough look at it first. One would also have to understand the relationships between the characteristics:
- Active: Active/Inactive
- TargetAirPurifierState: Manual/Auto
- CurrentAirPurifierState: Inactive/Idle/Purifying [my understandings: Inactive = Feedback state for inactive, Idle = Feedback state for active and auto mode (currently off), Purifying = Feedback state for active (Auto or manual, currently on)]
- optional: RotationSpeed
- optional: SwingMode
- optional: LockPhysicalControls

You could also add a linked Fan Service (with Active, CurrentFanState/TargetFanState/RotationDirection/RotationSpeed/SwingMode/LockPhysicalControls), but then changing "Active" in AirPurifier must result in a corresponding change in Fan. Changing "Active" in Fan is different: Fans can be activated exclusively (Fan-only), but deactivating Fan must also deactivate AirPurifier.

I think it shouldn't be made more complicated than necessary and one should only implement the characteristics that can actually be set via the home app. In addition, it might be useful not to link the MQTT topics 1:1 to Apple's characteristics, but to simplify them: One combined topic for Active and TargetAirPurifierState with modes: On, Auto, Off, (FanOnly)

Maybe I could take a closer look at the weekend. Otherwise we could continue to discuss here in the thread.

**Comment by EjvindHald** [comment]
Hi @tobekas 

Thank you very much for looking into this.

I would really, really appreciate a simple 1:1 transparent topics for Active, TargetAirPurifierState and for CurrentAirPurifierState with no logic on top. This will allow me 100 % control and flexibility +  full 2 way sync between my own device and Homekit. I am aware that the mqtt client will need more complexity, but that is fine as I also get the full benefits of the many nice features in the Homekit app. This full transparency is one of the really nice things with this mqtt plug-in.

On the other hand, I do recognize there is also a need for a simple easy to use approach as you write. Recently, @arachnetech has solved this very smart for the garage door - see #225 . He kept all the original topics 1:1 with Homekit and then on top of these added a new semi-automatic topic, which is easy to use. 

By following this approach you can both make this useful for the complex integration and the simple user. 

Thanks!

**Comment by tobekas** [comment]
Hi @EjvindHald,
I'd like a one-to-one mapping as well. Think this was also the basic idea of this project: to map the characteristics given by Apple directly via MQTT. The accessories should then follow these specifications if possible. Otherwise you have to add some logic yourself.
The only exception was the irrigation system, which requires several internal links and offers these timing functions.

Basically I can always recommend NodeRed if someone wants to implement some simple if-then logic, remapping of mqtt topics or small automations. For prototyping and playing around also great!

I would then keep it very simple at first.  Then I would see no problems at all. I may also check what other simple device types are missing. The majority is already implemented here.

**Comment by EjvindHald** [comment]
If you add the AirPurifier, then you probably need me to test it. I will be happy to do so, but I am not so familiar with how to install a branch version. Currently, I am running Homebridge in 2 instances in Docker, and I guess a smart way to test would be to initiate a third version in Docker based on a branch version. However, I do not know how to do that. Any input would be appreciated.

**Comment by tobekas** [comment]
Something like `npm install -g any-package@user-name/repo-name#branch-name` should do the job.

But it think after some testing from my side, it can take over directly into the main branch, since it will be very straight forward.

**Comment by arachnetech** [comment]
Thanks again, @tobekas. Incidentally, I completely agree with your sentiment:
 > I'd like a one-to-one mapping as well. Think this was also the basic idea of this project: to map the characteristics given by Apple directly via MQTT. The accessories should then follow these specifications if possible.

... and with NodeRed.

The recently-added codecs also make it easier to add user logic 'by configuration'. They're a bit more flexible than 'apply()' functions, as delivery of an incoming property change can be delayed/repeated.

Very happy to merge AirPurifier straight into the main branch once it's ready.

**Comment by EjvindHald** [comment]
I thinkI do not need any linked fan. My need is similar to [this ](https://www.youtube.com/watch?v=2mRc60crbN8)video (I do not know which country it is from). If possible, the filter options would be nice to have and of course the speed regulation. I guess the latter is RotationSpeed. Thanks.

**Comment by EjvindHald** [comment]
Hi @tobekas 
Are you expecting something from me or are you just busy with other stuff?
Thanks!

**Comment by tobekas** [comment]
Hi. Sorry, just busy... lot of work, sport and private issues... hardly time for hobby programming.

My next step is cleaning up PR #249 and then I could add AirPurifier.

**Comment by EjvindHald** [comment]
Take your time and your priorities - I get it when it is ready 👍

**Comment by EjvindHald** [comment]
> * Active: Active/Inactive
> * TargetAirPurifierState: Manual/Auto
> * CurrentAirPurifierState: Inactive/Idle/Purifying [my understandings: Inactive = Feedback state for inactive, Idle = Feedback state for active and auto mode (currently off), Purifying = Feedback state for active (Auto or manual, currently on)]
> * optional: RotationSpeed
> * optional: SwingMode
> * optional: LockPhysicalControls

I have tried to get insight into the options from Apple documentation. I agree with you on those you mentioned and I also agree that a linked Fan is not needed for now. Control of speed is essential, but I guess that is RotationSpeed as you mentioned.

If you have time, a nice to have feature includes those attributes from Apple documentation:
- FilterChangeIndication
- FilterLifeLevel
- FilterResetChangeIndication

They should all be optional.

Thanks.

**Comment by tobekas** [comment]
@EjvindHald: Finally I found time for implementing Air Purifier... :-)
I created a PR: #271
You could already test it from here: https://github.com/tobekas/homebridge-mqttthing/tree/AirPurifier but I think @arachnetech will merge it very soon.

**Comment by EjvindHald** [comment]
This is really, really great news :-) Looking forward to it, and I will test it as soon as it is merged!

Thanks again.

**Comment by tobekas** [comment]
Available in version 1.1.14

**Comment by EjvindHald** [comment]
I have now tested this, and so far I can confirm that everything works like a charm!
Only thing is to state in the readme description that getFilterChangeIndication is a Boolean. I also learned that this is the one to be used to reset the marking for filter replacement - simply set it to off, after the filter is replaced.
I am not quite sure, what is the idea behind ResetFilterIndication, which I can see is supported in Eve app. When the buttom is hit in the Eve app, I can see a mqtt message is sent, and then you could of course build logic around that.

**Comment by tobekas** [comment]
I'm glad you're pleased.

Exactly, `getFilterChangeIndication` and `setResetFilterIndication` are both Boolean topics (with the options to use integerValue, onValue, offValue, etc.). I also thought about implementing it with specific string values and making them configurable, like other state attributes. But would find it easier that way.

_ResetFilterIndication_ actually does not appear in the Home App, but in Eve. From the HomeKit point of view I suppose it is meant as follows: With _ResetFilterIndication_ the warning should be reset from within the app after a filter change. As a reaction _FilterChangeIndication_ should change back to false and also _FilterLifeTime_ should go to 100%. But the device itself must implement this (external logic). Without the reset topic this characteristic is not provided. It's optional. The device could also recognize the new filter itself and then reset _FilterChangeIndication_ automatically.  This logic would then always have to be provided by the MQTT device. Otherwise it works of course completely without the filter characteristics.

Can you make a suggestion how you think the readme should be completed? Gladly also via PullRequest...

Have fun with the new version!

**Comment by EjvindHald** [comment]
It is fine with that me that the getFilterChangeIndication is a boolean. However, if Apple decides to change this by adding a new string variant, it would be a breaking change. This is becuase then you need to change from a boolean to string, if it holds more than 2 values. But I am ok with current setup, and if you decide to change to a string, that will be fine as well.

Unfortunately, I am not able to do a PullRequest, but based on your suggestion, I will make an effort to set this up and learn it.

Instead I will try to make it as easy for you as possible.

Please replace this section:
Filter life level must be in the range 0 to 100 percent with no decimal places.

with this:
The filter life level is used to indicate remaining filter life level in percent. The value should be a integer between 0 and 100 with no decimal places.

Related to this is the filter change indication which is used to indicate if it is time to replace the filter. It is a boolean value and true indicates that a replacement is needed. After a filter replacement this should be set to false. 

If the hardware device does not have a way to indicate that the filter has been replaced, the reset filter indication can be used. It is currently only supported in the Eve app, and the idea is to click on it after a filter change, and then based on that execute a small piece of external logic which sets filter change indication to false.

**Comment by arachnetech** [comment]
Added, thanks.
